src/base.js

```diff
@@ -13,8 +13,225 @@
  * limitations under the License.
  */
 
+var Util = require('./util.js');
+
+// Start at a higher number to reduce chance of conflict
+var nextDisplayId = 1000;
+
+/**
+ * The base class for all VR displays.
+ */
+function VRDisplay() {
+  this.displayId = nextDisplayId++;
+  this.displayName = 'webvr-polyfill displayName';
+
+  this.isConnected = true;
+  this.isPresenting = false;
+  this.capabilities = {
+    hasPosition: false,
+    hasOrientation: false,
+    hasExternalDisplay: false,
+    canPresent: false
+  };
+  this.stageParameters = null;
+
+  // "Private" members.
+  this.waitingForPresent_ = false;
+  this.layer_ = null;
+
+  this.fullscreenEventTarget_ = null;
+  this.fullscreenChangeHandler_ = null;
+  this.fullscreenErrorHandler_ = null;
+}
+
+VRDisplay.prototype.requestAnimationFrame = function(callback) {
+  return window.requestAnimationFrame(callback);
+};
+
+VRDisplay.prototype.cancelAnimationFrame = function(id) {
+  return window.cancelAnimationFrame(id);
+};
+
+VRDisplay.prototype.requestPresent = function(layer) {
+  var self = this;
+  this.layer_ = layer;
+
+  return new Promise(function(resolve, reject) {
+    if (!self.capabilities.canPresent) {
+      reject(new Error("VRDisplay is not capable of presenting."));
+      return;
+    }
+
+    self.waitingForPresent_ = false;
+    if (layer && layer.source) {
+      function onFullscreenChange() {
+        var fullscreenElement = document.fullscreenElement ||
+            document.webkitFullscreenElement ||
+            document.mozFullScreenElement ||
+            document.msFullscreenElement;
+
+        self.isPresenting = fullscreenElement == layer.source;
+        self.fireVRDisplayPresentChange_();
+        if (self.isPresenting) {
+          if (screen.orientation && screen.orientation.lock)
+            screen.orientation.lock("landscape-primary");
+          self.waitingForPresent_ = false;
+          self.beginPresent_();
+          resolve();
+        } else {
+          if (screen.orientation && screen.orientation.unlock)
+            screen.orientation.unlock();
+          self.endPresent_();
+          self.removeFullscreenListeners_();
+        }
+      }
+      function onFullscreenError() {
+        if (!self.waitingForPresent_)
+          return;
+
+        self.removeFullscreenListeners_();
+
+        self.waitingForPresent_ = false;
+        self.isPresenting = false;
+
+        reject(new Error("Unable to present."));
+      }
+
+      self.addFullscreenListeners_(layer.source,
+          onFullscreenChange, onFullscreenError);
+
+      self.waitingForPresent_ = true;
+      // TODO: Spec says that requestFullscreen returns a promise, but that
```

Comment:
yeah, the Promise behaviour described in the WHATWG Fullscreen spec hasn't shipped anywhere yet